components/builder-db/src/async.rs

```diff
@@ -0,0 +1,226 @@
+        let has_key = {
+            let d = self.dispatch
+                .read()
+                .expect("Async dispatch lock is poisoned");
+            d.contains_key(dispatch_key)
+        };
+        let is_scheduled = {
+            let r = self.retry.read().expect("Async retry lock is poisoned");
+            r.contains_key(dispatch_key)
+        };
+        if has_key && !is_scheduled {
+            let function = {
+                let d = self.dispatch
+                    .read()
+                    .expect("Async dispatch lock is poisoned");
+                // Safe because we checked above
+                d.get(dispatch_key).unwrap().clone()
+            };
+            let mut r = self.retry
+                .write()
+                .expect("Async retry lock is poisoned");
+            r.insert(dispatch_key.to_string(), (SteadyTime::now(), function));
+        }
+        Ok(())
+    }
+
+    pub fn register(&self, dispatch_key: DispatchKey, callback: EventFunction) {
+        {
+            let mut d = self.dispatch
+                .write()
+                .expect("Async dispatch lock is poisoned");
+            d.insert(dispatch_key.clone(), callback.clone());
+        }
+        {
+            let mut r = self.retry
+                .write()
+                .expect("Async retry lock is poisoned");
+            r.insert(dispatch_key, (SteadyTime::now(), callback));
+        }
+    }
+
+    fn get_num_events(&self, available_jobs: usize) -> Vec<(DispatchKey, EventFunction)> {
+        let r = self.retry.read().expect("Async retry lock is poisoned");
+        r.iter()
+            .filter(|&(_dispatch_key, &(event_time, _callback))| SteadyTime::now() >= event_time)
+            .take(available_jobs)
+            .map(|(dispatch_key, &(_event_time, callback))| {
+                     (dispatch_key.to_string(), callback.clone())
+                 })
+            .collect()
+    }
+
+    pub fn run_event(&self, key: DispatchKey, event: EventFunction) {
+        let remove_key = key.clone();
+        match event(self.pool.clone()) {
+            Ok(EventOutcome::Finished) => {
+                debug!("Event finished {}", key);
+                {
+                    let mut f = self.failure_count
+                        .write()
+                        .expect("Async failure count lock poisoned");
+                    f.remove(&key);
+                }
+            }
+            Ok(EventOutcome::Retry) => {
+                self.retry_failed_event(key, event);
+            }
+            Err(e) => {
+                error!("Event {} failed, {}", key, e);
+                self.retry_failed_event(key, event);
+            }
+        }
+        let mut running = self.running.write().expect("Running lock is poisoned");
+        running.remove(&remove_key);
+    }
+
+    fn retry_failed_event(&self, key: DispatchKey, event: EventFunction) {
+        error!("Scheduling retry of {:?}", key);
```

Comment:
We should reserve `error` level debugging for the case where we want to be woken up in the middle of the night and `warn` level when we want to see that it happened in the morning. I think this is a case for `warn` here